gtk4/src/subclass/tree_model.rs

```diff
@@ -0,0 +1,503 @@
+
+            if let Some(f) = (*iface).row_inserted.as_ref() {
+                f(
+                    tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0,
+                    path.to_glib_none().0 as *mut _,
+                    iter.to_glib_none().0 as *mut _,
+                );
+            }
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+        }
+    }
+
+    fn row_has_child_toggled(&self, tree_model: &Self::Type, path: &TreePath, iter: &TreeIter) {
+        unsafe {
+            let type_ = ffi::gtk_print_operation_preview_get_type();
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+
+            if let Some(f) = (*iface).row_has_child_toggled.as_ref() {
+                f(
+                    tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0,
+                    path.to_glib_none().0 as *mut _,
+                    iter.to_glib_none().0 as *mut _,
+                );
+            }
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+        }
+    }
+
+    fn row_deleted(&self, tree_model: &Self::Type, path: &TreePath) {
+        unsafe {
+            let type_ = ffi::gtk_print_operation_preview_get_type();
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+
+            if let Some(f) = (*iface).row_deleted.as_ref() {
+                f(
+                    tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0,
+                    path.to_glib_none().0 as *mut _,
+                );
+            }
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+        }
+    }
+
+    fn get_flags(&self, tree_model: &Self::Type) -> TreeModelFlags {
+        unsafe {
+            let type_ = ffi::gtk_print_operation_preview_get_type();
+            let iface = glib::gobject_ffi::g_type_default_interface_ref(type_)
+                as *mut ffi::GtkTreeModelIface;
+            assert!(!iface.is_null());
+
+            let f = (*iface).get_flags.as_ref().unwrap();
+            let ret = f(tree_model.unsafe_cast_ref::<TreeModel>().to_glib_none().0);
+
+            glib::gobject_ffi::g_type_default_interface_unref(iface as glib::ffi::gpointer);
+            from_glib(ret)
+        }
+    }
+
+    fn get_n_columns(&self, tree_model: &Self::Type) -> i32;
+    fn get_column_type(&self, tree_model: &Self::Type, index: i32) -> glib::Type;
+    fn get_iter(&self, tree_model: &Self::Type, path: &TreePath) -> Option<TreeIter>;
+    fn get_path(&self, tree_model: &Self::Type, iter: &TreeIter) -> TreePath;
+    fn get_value(&self, tree_model: &Self::Type, iter: &TreeIter, index: i32) -> Value;
+    fn iter_next(&self, tree_model: &Self::Type) -> Option<TreeIter>;
```

Comment:
Can we actually implement all these functions? Do we have ways to create a tree iter out of thin air?

Reply:
We don't actually have a way to create a TreeIter currently no, but it's just a sturct around an int & 3 random user_data pointers; I will have to look at how other bindings handle this (if they do)

Reply:
So none of the bindings provides a way to create a TreeIter but as we made the whole trait unsafe, I guess that's okay as is.

Reply:
I'd rather not have this trait for now then. It clearly shows the it's not usable yet and not enough thought was put into how to make it work in Rust. Having the trait around in an unusable form will just confuse users that try to use it.

Maybe move this into a separate WIP PR?

Reply:
Sure, I can move the FontChooser out of this PR as the rest is dependant on TreeModel. It could definitely use some better approach to make it usable from Rust but requires someone who is actually familiar with the Tree* stuff in GTK :) 
Also I couldn't find that many apps that has custom TreeModels, most of them are written in C like sysprof.